### PR TITLE
LPS-175152 Accept a leading slash in redirection patterns

### PR DESCRIPTION
**Pull request by adolfopa**
# What is this about?

https://issues.liferay.com/browse/LPS-175152

Redirection patterns will not match with a leading slash, but replacement patterns must begin with one.  Ignore the first leading slash in the pattern to make matching a bit less confusing.

### CODE REVIEW

**Comment by liferay-lima** [comment]
ci:test:sf

File: modules/apps/redirect/redirect-service/src/main/java/com/liferay/redirect/internal/util/PatternUtil.java [hunk 1]
```diff
@@ -45,10 +45,20 @@ public static Map<Pattern, String> parse(String[] patternStrings) {
...
+
 			return patternString;
 		}
 
+		if (patternString.startsWith(StringPool.SLASH)) {
+			patternString = patternString.substring(1);
+		}
+
 		return StringPool.CARET + patternString;
 	}
```
Comment (marcogalluzzi) [inline modules/apps/redirect/redirect-service/src/main/java/com/liferay/redirect/internal/util/PatternUtil.java:60]: Maybe you find this version easier to understand:

```
	private static String _normalize(String patternString) {
		if (patternString.startsWith(StringPool.CARET)) {
			patternString = patternString.substring(1);
		}

		if (patternString.startsWith(StringPool.SLASH)) {
			patternString = patternString.substring(1);
		}

		return StringPool.CARET + patternString;
	}
```

File: modules/apps/redirect/redirect-service/src/test/java/com/liferay/redirect/internal/util/PatternUtilTest.java [hunk 2]
```diff
@@ -48,6 +48,16 @@ public void testAnchoredPattern() {
 			patternStrings.toString(), 1, patternStrings.size());
 	}
 
+	@Test
+	public void testCaretSlashPattern() {
+		Map<Pattern, String> patternStrings = PatternUtil.parse(
+			new String[] {"^/xyz abc"});
+
+		Assert.assertEquals("^xyz", _getFistPatternString(patternStrings));
```
Comment (marcogalluzzi) [inline modules/apps/redirect/redirect-service/src/test/java/com/liferay/redirect/internal/util/PatternUtilTest.java:56]: Shouldn't this method be called `_getFirstPatternString` ? (not _getFist...)

File: modules/apps/redirect/redirect-service/src/test/java/com/liferay/redirect/internal/util/PatternUtilTest.java [hunk 3]
```diff
@@ -68,6 +78,16 @@ public void testInvalidRegexPattern() {
 		PatternUtil.parse(new String[] {"*** a"});
 	}
 
+	@Test
+	public void testSlashPattern() {
```
Comment (marcogalluzzi) [inline modules/apps/redirect/redirect-service/src/test/java/com/liferay/redirect/internal/util/PatternUtilTest.java:82]: Would it be interesting to have also a test called `testCaretPattern()`?

File: modules/apps/redirect/redirect-service/src/main/java/com/liferay/redirect/internal/util/PatternUtil.java (hunk 1 above)
Comment (adolfopa) [reply modules/apps/redirect/redirect-service/src/main/java/com/liferay/redirect/internal/util/PatternUtil.java]: Yes, I agree.

File: modules/apps/redirect/redirect-service/src/test/java/com/liferay/redirect/internal/util/PatternUtilTest.java (hunk 2 above)
Comment (adolfopa) [reply modules/apps/redirect/redirect-service/src/test/java/com/liferay/redirect/internal/util/PatternUtilTest.java]: Yes, that was a typo.

File: modules/apps/redirect/redirect-service/src/test/java/com/liferay/redirect/internal/util/PatternUtilTest.java (hunk 3 above)
Comment (adolfopa) [reply modules/apps/redirect/redirect-service/src/test/java/com/liferay/redirect/internal/util/PatternUtilTest.java:92]: Sure.

**Comment by adolfopa** [comment]
ci:test:sf

**Comment by adolfopa** [comment]
ci:test:relevant

**Review by marcogalluzzi — APPROVED** [review]

**Comment by marcogalluzzi** [comment]
ci:forward

**Comment by marcogalluzzi** [comment]
Hi @adolfopa it looks there is a semantic versioning error to solve:
```
     [exec]       +   access     final
     [exec]       +   access     static
     [exec]       +   constant   TERMS_OF_USE
     [exec]       +   return     java.lang.String
     [exec]     +   method     <init>()
     [exec]       +   return     void
     [exec]     +   method     clone()
     [exec]       +   access     protected
     [exec]       +   return     java.lang.Object
     [exec]     +   method     equals(java.lang.Object)
     [exec]       +   return     boolean
     [exec]     +   method     finalize()
     [exec]       +   access     protected
     [exec]       +   return     void
     [exec]     +   method     hashCode()
     [exec]       +   return     int
     [exec]     +   method     toString()
     [exec]       +   return     java.lang.String
     [exec]   -   class      com.liferay.layout.utility.page.kernel.constants.LayoutUtilityPageEntryTypesConstants
     [exec]     -   field      LAYOUT
     [exec]       -   access     final
     [exec]       -   access     static
     [exec]       -   constant   LAYOUT
     [exec]       -   return     java.lang.String
     [exec]     -   field      STATUS
     [exec]       -   access     final
     [exec]       -   access     static
     [exec]       -   constant   STATUS
     [exec]       -   return     java.lang.String
     [exec]     -   field      TERMS_OF_USE
     [exec]       -   access     final
     [exec]       -   access     static
     [exec]       -   constant   TERMS_OF_USE
     [exec]       -   return     java.lang.String
     [exec]     -   method     <init>()
     [exec]       -   return     void
     [exec]     -   method     clone()
     [exec]       -   access     protected
     [exec]       -   return     java.lang.Object
     [exec]     -   method     equals(java.lang.Object)
     [exec]       -   return     boolean
     [exec]     -   method     finalize()
     [exec]       -   access     protected
     [exec]       -   return     void
     [exec]     -   method     hashCode()
     [exec]       -   return     int
     [exec]     -   method     toString()
     [exec]       -   return     java.lang.String
     [exec]   -   version    1.0.0
     [exec]   +   version    1.1.0
     [exec] [Baseline Report] Mode: diff (persisted)
     [exec] Semantic versioning is incorrect while checking /opt/dev/projects/github/liferay-portal/portal-kernel/portal-kernel.jar against /opt/dev/projects/github/liferay-portal/.gradle/caches/modules-2/files-2.1/com.liferay.portal/com.liferay.portal.kernel/96.0.1/87d20d3daea0984b5ad25b45023aaa1ee46c99ec/com.liferay.portal.kernel-96.0.1.jar
```

**Comment by adolfopa** [comment]
That has nothing to do with these changes. Needs to be fixed upstream.

**Comment by Jeremy-Ch** [comment]
ci:forward

**Comment by adolfopa** [comment]
ci:forward

**Pull request closed without merging**